Cargamos las librerias que utilizaremos

In [1]:
#Importamos las librerias:
import pandas as pd
import numpy as np
# Módulo para manipular paths
from pathlib import Path
# Librerí para identificar el encoding de los archivos 
import chardet
# Módulo para manipular expresiones regulares 
import re
import os
import json

Importamos las librerias que usaremos para conectar y subir con mysql:

In [2]:
import pymysql
from sqlalchemy import create_engine

Extraccion:
Los archivos tienen diversos formatos y argumentos de importación. Para automatizarlo, creamos una función que considere todas estas consideraciones para instanciarlo con el path de cada file.

In [3]:
#Creamos una funcion para cargar los datos
def importar(list):
    
    
    dirPath = r"./Datasets"
    
    for path in list: 
        #concatenamos el path para que nos tome bien la ruta
            path= os.path.join(dirPath, path)
        
            if Path(path).suffix == ".csv":            
                f = open(path,"rb")
                result= chardet.detect(f.read())

                df= pd.read_csv(path,encoding=result["encoding"], sep=None, engine="python",decimal=".")

            if Path(path).suffix==".json"or Path(path).suffix==".js":

                df=pd.read_json(path, precise_float=True)
    
            return df

In [4]:
#Buscar la ruta de los archivos
dirPath = r"./Datasets"
list_amazon=[]
list_disney=[]
list_hulu=[]
list_netflix=[]
for f in os.listdir(dirPath):
    if os.path.isfile(os.path.join(dirPath, f)):
        if "amazon" in f:
            list_amazon.append(f)
        elif "disney" in f:
            list_disney.append(f)
        elif "hulu" in f:
            list_hulu.append(f)
        elif "netflix" in f:
            list_netflix.append(f)
            

Traemos los archivos transformados en DF

In [5]:
#importamos amazon titles
amazon = importar(list_amazon)

In [6]:
#importamos disney titles
disney = importar(list_disney)

In [7]:
#importar hulu titles
hulu = importar(list_hulu)

In [8]:
#importamos netflix titles
netflix = importar(list_netflix)

Vamos a crear una columna con el origen de cada archivo asi podemos identificarlo en el tiempo


In [9]:
#creamos una columna nueva en cada tabla con su nombre de origen
amazon["platform"]= "amazon"
disney["platform"]= "disney"
hulu["platform"]= "hulu"
netflix["platform"]= "netflix"

Ahora uniremos todos los df en uno solo para poder seguir limpiandolo

In [10]:
#creamos nuestro df que usaremos como principal
df_stream = amazon

In [11]:
#concatenamos el resto de df
df_stream = pd.concat([df_stream,disney])
df_stream = pd.concat([df_stream,hulu])
df_stream = pd.concat([df_stream,netflix])

In [14]:
df_stream.sample(5)

show_id     type                                    title  \
1611   s1612  TV Show                          Are You The One   
1528   s1529  TV Show  Napoleon: The Man Who Would Rule Europe   
2147   s2148    Movie                      Diamonds in the Sky   
1550   s1551    Movie                             Danger Close   
473     s474    Movie                             Born to Play   

                 director                                               cast  \
1611                 None                                        Ryan Devlin   
1528                  NaN                                                NaN   
2147       Kunle Afolayan  Joke Silva, Omowunmi Dada, Bimbo Akintola, Fem...   
1550                  NaN                                                NaN   
473   Viridiana Lieberman                                               None   

            country        date_added  release_year  rating   duration  \
1611  United States  December 1, 2020          2014   TV-14  2 Seasons   
1528            NaN               NaN          2019     13+   1 Season   
2147           None    August 8, 2020          2018   TV-14    113 min   
1550  United States      July 1, 2020          2017  86 min        NaN   
473   United States     July 12, 2021          2020   TV-MA     93 min   

                          listed_in  \
1611                     Reality TV   
1528  Documentary, Special Interest   
2147   Dramas, International Movies   
1550                  Documentaries   
473    Documentaries, Sports Movies   

                                            description platform  
1611  Determined by a dating algorithm, a group of s...  netflix  
1528  It is the first full series appraisal of the l...   amazon  
2147  Three families navigate uncertainty and strugg...  netflix  
1550  War reporter Alex Quade re-embeds with Green B...     hulu  
473   This documentary follows the Boston Renegades ...  netflix

Habiendo hecho varios samples para ir verificando los datos, me encontre que datos que pertenecen a la columna"duration" se encuentran en la columna "ranting".   
Se procede a cambiarlos de ubicacion para no perderlos

In [15]:
df_stream.isnull().sum()

show_id             0
type                0
title               0
director         8259
cast             5321
country         11499
date_added       9554
release_year        0
rating            864
duration          482
listed_in           0
description         4
platform            0
dtype: int64

In [16]:
df_stream['duration'] = df_stream.apply(lambda row: row['rating'] if ((pd.isnull(row['duration'])) & (("min" in str(row['rating'])) or ("Season" in str(row['rating'])))) else row['duration'], axis=1)


In [17]:
df_stream.isnull().sum()

show_id             0
type                0
title               0
director         8259
cast             5321
country         11499
date_added       9554
release_year        0
rating            864
duration          187
listed_in           0
description         4
platform            0
dtype: int64

Como se puede apreciar los nulos de "duration" disminuyeron de 482 a 187 seria que disminuyo un 61.2% la cantidad de nulos

In [15]:
#controlamos el porcentaje de nulos en nuestras columnas
round(df_stream.isna().mean()*100,2)

show_id          0.00
type             0.00
title            0.00
director        35.91
cast            23.14
country         50.00
date_added      41.54
release_year     0.00
rating           3.76
duration         0.81
listed_in        0.00
description      0.02
platform         0.00
dtype: float64

Columnas necesarias para realizar las consultas pedidas por el cliente:  
- platform (se va a crear esta columna)  
- type (cambiar nombre a "category" para que no choque con mysql)  
- duration  
- listed_in(cambiar nombre a "genre")  
- release_year  
- cast 

Columnas que vamos a dejar para futuras consultas:  
- title     
---------------------------------------------------------------------------------------------------------------------------------------

Las siguientes columnas van a ser dropeadas del df por tener una gran cantidad de nulos y no ser relevantes:  
- director: 35.91% nulos  
- date_added: 41.54% nulos    
- country: 50% nulos   

Columnas que vamos a dropear por no ser relevantes en las consultas que deseamos realizar:
- show_id  
- rating  
- description 


In [18]:
#Dropeamos las columnas que decidimos:
df_stream = df_stream.drop(["show_id","director","country","date_added","rating","description"],axis=1)

In [19]:
df_stream.head(1)

type                title                                            cast  \
0  Movie  The Grand Seduction  Brendan Gleeson, Taylor Kitsch, Gordon Pinsent   

   release_year duration      listed_in platform  
0          2014  113 min  Comedy, Drama   amazon

In [20]:
#borramos los indices originales de cada df
df_stream = df_stream.reset_index(drop=True)

Ya que necesitamos usar la columna duration pasa saber valores numericos, necesitamos separar los numeros de la medida de tiempo en 2 columnas

In [21]:
#Creamos un df nuevo con 2 columnas, una con el numero y otra con el tipo de tiempo
separar= df_stream.duration.str.split(" ", n=1, expand=True)

Creamos 2 columnas con los datos separados

In [22]:
df_stream["duration"] = separar[0]

In [24]:
df_stream["type_duration"] = separar[1]

In [26]:
df_stream

type                 title  \
0        Movie   The Grand Seduction   
1        Movie  Take Care Good Night   
2        Movie  Secrets of Deception   
3        Movie    Pink: Staying True   
4        Movie         Monster Maker   
...        ...                   ...   
22993    Movie                Zodiac   
22994  TV Show           Zombie Dumb   
22995    Movie            Zombieland   
22996    Movie                  Zoom   
22997    Movie                Zubaan   

                                                    cast  release_year  \
0         Brendan Gleeson, Taylor Kitsch, Gordon Pinsent          2014   
1       Mahesh Manjrekar, Abhay Mahajan, Sachin Khedekar          2018   
2      Tom Sizemore, Lorenzo Lamas, Robert LaSardo, R...          2017   
3      Interviews with: Pink, Adele, Beyoncé, Britney...          2014   
4      Harry Dean Stanton, Kieran O'Brien, George Cos...          1989   
...                                                  ...           ...   
22993  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...          2007   
22994                                               None          2018   
22995  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...          2009   
22996  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...          2006   
22997  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          2015   

      duration                                       listed_in platform  \
0          113                                   Comedy, Drama   amazon   
1          110                            Drama, International   amazon   
2           74                         Action, Drama, Suspense   amazon   
3           69                                     Documentary   amazon   
4           45                                  Drama, Fantasy   amazon   
...        ...                                             ...      ...   
22993      158                  Cult Movies, Dramas, Thrillers  netflix   
22994        2          Kids' TV, Korean TV Shows, TV Comedies  netflix   
22995       88                         Comedies, Horror Movies  netflix   
22996       88              Children & Family Movies, Comedies  netflix   
22997      111  Dramas, International Movies, Music & Musicals  netflix   

      type_duration  
0               min  
1               min  
2               min  
3               min  
4               min  
...             ...  
22993           min  
22994       Seasons  
22995           min  
22996           min  
22997           min  

[22998 rows x 8 columns]

In [20]:
df_stream.sample(4)

type                        title  \
8385     Movie                    Max Payne   
11461  TV Show                       Gossip   
11058    Movie              Tru Confessions   
19147    Movie  Phir Bhi Dil Hai Hindustani   

                                                    cast  release_year  \
8385   Mark Wahlberg, Mila Kunis, Beau Bridges, Ludac...          2008   
11461                                                NaN          2021   
11058  Shia LaBeouf, Clara Bryant, William McGuire, N...          2002   
19147  Shah Rukh Khan, Juhi Chawla, Paresh Rawal, Sat...          2000   

      duration                               listed_in platform type_duration  
8385       100                                  Action   amazon           min  
11461        1                           Documentaries     hulu        Season  
11058       83                    Coming of Age, Drama   disney           min  
19147      159  Comedies, Dramas, International Movies  netflix           min

Verificamos los nulos que quedan en nuestro df_stream

In [24]:
df_stream.isnull().sum()

type                0
title               0
cast             5321
release_year        0
duration          187
listed_in           0
platform            0
type_duration     187
dtype: int64

In [27]:
#remplazamos nulos por "sin dato" en todo el df
df_stream = df_stream.fillna("sin dato")

In [26]:
#Verificamos si hay datos duplicados
duplicado = df_stream[df_stream.duplicated()==True].count()
print(duplicado)

type             0
title            0
cast             0
release_year     0
duration         0
listed_in        0
platform         0
type_duration    0
dtype: int64


Al ver que duration es un type object y tendria que ser de tipo numerico para realizar las consultas, lo modificamos.  
Se avisa al cliente que los datos nulos fueron cambiados por 0 en la columna "duration"

In [28]:
#con la siguiente sentencia lo pasamos a float y todos los "sin datos" pasaran a NaN
df_stream["duration"] = pd.to_numeric(df_stream["duration"], errors="coerce")
#remplazamos los nulos por 0
df_stream["duration"] = df_stream["duration"].fillna(0)
#cambiamos de formato float a int
df_stream["duration"]=df_stream["duration"].astype(int)

Creamos un nuevo index

In [39]:
#creamos un id para cada fila
df_stream["idStream"] = df_stream.index

In [29]:
df_stream

type                 title  \
0        Movie   The Grand Seduction   
1        Movie  Take Care Good Night   
2        Movie  Secrets of Deception   
3        Movie    Pink: Staying True   
4        Movie         Monster Maker   
...        ...                   ...   
22993    Movie                Zodiac   
22994  TV Show           Zombie Dumb   
22995    Movie            Zombieland   
22996    Movie                  Zoom   
22997    Movie                Zubaan   

                                                    cast  release_year  \
0         Brendan Gleeson, Taylor Kitsch, Gordon Pinsent          2014   
1       Mahesh Manjrekar, Abhay Mahajan, Sachin Khedekar          2018   
2      Tom Sizemore, Lorenzo Lamas, Robert LaSardo, R...          2017   
3      Interviews with: Pink, Adele, Beyoncé, Britney...          2014   
4      Harry Dean Stanton, Kieran O'Brien, George Cos...          1989   
...                                                  ...           ...   
22993  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...          2007   
22994                                           sin dato          2018   
22995  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...          2009   
22996  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...          2006   
22997  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          2015   

       duration                                       listed_in platform  \
0           113                                   Comedy, Drama   amazon   
1           110                            Drama, International   amazon   
2            74                         Action, Drama, Suspense   amazon   
3            69                                     Documentary   amazon   
4            45                                  Drama, Fantasy   amazon   
...         ...                                             ...      ...   
22993       158                  Cult Movies, Dramas, Thrillers  netflix   
22994         2          Kids' TV, Korean TV Shows, TV Comedies  netflix   
22995        88                         Comedies, Horror Movies  netflix   
22996        88              Children & Family Movies, Comedies  netflix   
22997       111  Dramas, International Movies, Music & Musicals  netflix   

      type_duration  idStream  
0               min         0  
1               min         1  
2               min         2  
3               min         3  
4               min         4  
...             ...       ...  
22993           min     22993  
22994       Seasons     22994  
22995           min     22995  
22996           min     22996  
22997           min     22997  

[22998 rows x 9 columns]

In [40]:
#re-organizamos las columnas
df_stream= df_stream.reindex(columns=["idStream","type","title","cast","release_year","duration","type_duration","listed_in","platform"])

In [41]:
#quitamos los espacios en blanco en las siguientes columnas:
df_stream["type"] = df_stream["type"].str.strip()
df_stream["title"] = df_stream["title"].str.strip()
df_stream["type_duration"] = df_stream["type_duration"].str.strip()

Separamos los dos df nuevas de "actores" y "genero" para hacer futuras consultas en sql

**actores**

In [42]:
#generar el nuevo df y borramos duplicados
df_actor = df_stream[["idStream","cast"]].drop_duplicates()

In [43]:
#creamos una lista con nuestros valores separados por coma
df_actor["cast"] = df_actor["cast"].str.split(",")
#separamos estos valores en una columna, cada uno con su valor de idStream
df_actor = df_actor.explode("cast")

In [44]:
#sacamos los espacios en blanco que tenga nuestra columna
df_actor["cast"] = df_actor["cast"].str.strip()

**genero**

In [45]:
#generar el nuevo df y borramos duplicados
df_genero = df_stream[["idStream","listed_in"]].drop_duplicates()

In [46]:
#creamos una lista con nuestros valores separados por coma
df_genero["listed_in"] = df_genero["listed_in"].str.split(",")
#separamos estos valores en una columna, cada uno con su valor de idStream
df_genero = df_genero.explode("listed_in")

In [47]:
#sacamos los espacios en blanco que tenga nuestra columna
df_genero["listed_in"]= df_genero["listed_in"].str.strip()

In [48]:
df_genero

idStream                 listed_in
0             0                    Comedy
0             0                     Drama
1             1                     Drama
1             1             International
2             2                    Action
...         ...                       ...
22996     22996  Children & Family Movies
22996     22996                  Comedies
22997     22997                    Dramas
22997     22997      International Movies
22997     22997          Music & Musicals

[48303 rows x 2 columns]

In [39]:
df_actor

idStream                   cast
0             0        Brendan Gleeson
0             0          Taylor Kitsch
0             0         Gordon Pinsent
1             1       Mahesh Manjrekar
1             1          Abhay Mahajan
...         ...                    ...
22997     22997       Manish Chaudhary
22997     22997           Meghna Malik
22997     22997          Malkeet Rauni
22997     22997         Anita Shabdish
22997     22997  Chittaranjan Tripathy

[119713 rows x 2 columns]

Cambiamos los nombres de las columnas antes de insertarlas en el db

In [49]:
#modificamos el nombre "listed_in" a "genre"
df_genero.rename(columns={"listed_in":"genre"},inplace=True)
df_stream.rename(columns={"type":"category"},inplace=True)

Ya que creamos los dos nuevos df, vamos a dropear estas columnas de nuestro df_stream

In [41]:
df_stream.head(1)

idStream category                title  \
0         0    Movie  The Grand Seduction   

                                             cast  release_year  duration  \
0  Brendan Gleeson, Taylor Kitsch, Gordon Pinsent          2014       113   

  type_duration      listed_in platform  
0           min  Comedy, Drama   amazon

In [50]:
#dropear columnas inecesarias
df_stream = df_stream.drop(["cast","listed_in"], axis=1)

In [43]:
df_stream.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22998 entries, 0 to 22997
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   idStream       22998 non-null  int64 
 1   category       22998 non-null  object
 2   title          22998 non-null  object
 3   release_year   22998 non-null  int64 
 4   duration       22998 non-null  int32 
 5   type_duration  22998 non-null  object
 6   platform       22998 non-null  object
dtypes: int32(1), int64(2), object(4)
memory usage: 1.1+ MB


In [44]:
df_genero.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48303 entries, 0 to 22997
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   idStream  48303 non-null  int64 
 1   genre     48303 non-null  object
dtypes: int64(1), object(1)
memory usage: 1.1+ MB


In [45]:
df_actor.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 119713 entries, 0 to 22997
Data columns (total 2 columns):
 #   Column    Non-Null Count   Dtype 
---  ------    --------------   ----- 
 0   idStream  119713 non-null  int64 
 1   cast      119713 non-null  object
dtypes: int64(1), object(1)
memory usage: 2.7+ MB


In [53]:
df_genero.genre.unique()

array(['Comedy', 'Drama', 'International', 'Action', 'Suspense',
       'Documentary', 'Fantasy', 'Kids', 'Special Interest',
       'Science Fiction', 'Adventure', 'Horror', 'Sports',
       'Talk Show and Variety', 'Anime', 'Arts', 'Entertainment',
       'and Culture', 'TV Shows', 'Animation',
       'Music Videos and Concerts', 'Fitness', 'Faith and Spirituality',
       'Military and War', 'Western', 'LGBTQ', 'Romance', 'Unscripted',
       'Young Adult Audience', 'Arthouse', 'Historical', 'Family',
       'Musical', 'Docuseries', 'Music', 'Biographical',
       'Action-Adventure', 'Superhero', 'Reality', 'Survival',
       'Animals & Nature', 'Coming of Age', 'Lifestyle', 'Movies',
       'Concert Film', 'Crime', 'Anthology', 'Medical', 'Variety',
       'Spy/Espionage', 'Buddy', 'Parody', 'Game Show / Competition',
       'Romantic Comedy', 'Thriller', 'Police/Cop', 'Talk Show', 'Dance',
       'Series', 'Mystery', 'Soap Opera / Melodrama', 'Disaster',
       'Travel', 'Stand Up

In [54]:
df_stream.sample()

idStream   type      title  release_year  duration type_duration  \
7538      7539  Movie  Route 666          2001        90           min   

     platform  
7538   amazon

In [57]:
df_stream

idStream category                 title  release_year  duration  \
0             0    Movie   The Grand Seduction          2014       113   
1             1    Movie  Take Care Good Night          2018       110   
2             2    Movie  Secrets of Deception          2017        74   
3             3    Movie    Pink: Staying True          2014        69   
4             4    Movie         Monster Maker          1989        45   
...         ...      ...                   ...           ...       ...   
22993     22993    Movie                Zodiac          2007       158   
22994     22994  TV Show           Zombie Dumb          2018         2   
22995     22995    Movie            Zombieland          2009        88   
22996     22996    Movie                  Zoom          2006        88   
22997     22997    Movie                Zubaan          2015       111   

      type_duration platform  
0               min   amazon  
1               min   amazon  
2               min   amazon  
3               min   amazon  
4               min   amazon  
...             ...      ...  
22993           min  netflix  
22994       Seasons  netflix  
22995           min  netflix  
22996           min  netflix  
22997           min  netflix  

[22998 rows x 7 columns]

**CARGAR A LA BASE DE DATOS**

In [57]:
#creamos nuestro engine con la ruta de nuestra db
engine = create_engine("mysql+pymysql://root:1234@localhost/pi_henry_etl")

In [58]:
#cargamos los df por medio de sqlalchemy
df_stream.to_sql("streaming",con=engine,index=False,if_exists='append')

22998

In [54]:
df_genero.to_sql("genre_table",con=engine,index=False,if_exists='append')

48303

In [55]:
df_actor.to_sql("cast_table",con=engine,index=False,if_exists='append')

119713